### ISSUE TITLE
Watcher options

### ISSUE THREAD

**Issue by kelunik** [async-interop/event-loop#12]
In Amp we currently have two watcher options:
- `keep_alive`<br>Our event loop will exit once no more watchers are active. If `keep_alive` is set to false, the watcher won't count against the "active watcher count".
- `enable`<br>Watchers are enabled by default when created. This options allows to create disabled watchers, e.g. write watchers that don't need to be enabled yet.

Do we want to integrate both like it's done in Amp via an `$options` array?

**Comment by bwoebi** [comment]
Note about the keep_alive option:
I think it is necessary because of timers and signal watchers mainly, which you need during the whole loop run.
E.g. a timer which calculates `date("r")` every second to save an expensive function call. (and `__destruct` obviously not working because the loop is referencing the object)
Or signal watchers in general.

This is most important for the case when the loop needs to ever end (we don't want zombie processes ;-)) like a cronjob for example.

`enable` is not so important and just a comfort function to not require an immediate disable call. So, while we may add it, we don't _need_ to.

**Comment by AndrewCarterUK** [comment]
Array options are something I'd like to avoid if at all possible.

It's can be difficult to detect typos when testing and I think it has the potential to encourage mistakes.

I've had to handle situations before where faults such as the typo below were not detected in unit tests.

``` php
EventLoop::onReadable($stream, $callback, ['keep_aliv' => true]);
```

I agree that `enable` is a comfort function that I think we can live with out.

Regarding the `keep_alive` parameter, I think we should, if we can, think of another way to accommodate for this use case without adding an array options method parameter.

Could a method that exposed the active watchers be a suitable replacement?

``` php
interface EventLoop
{
    // ...

    /**
     * Returns an array of active watcher identifiers.
     *
     * @param $disabled Set to true to return disabled AND enabled watchers.
     * @return string[]
     */
    public function getWatchers($disabled = false);
}
```

How about an `onEmpty()` event that was fired on loop ticks with no watchers? That would allow for:

``` php
EventLoop::onEmpty(function () {
    EventLoop::stop();
});
```

I'm just throwing out ideas here, but the React and Icicle event loops have avoided the need for this parameter - so I'd rather omit it if possible!

**Comment by bwoebi** [comment]
Array options are the cleanest way though if we ever want to have more options. (Or perhaps need to have loop-specific options when talking to a concrete implementation).

Regarding typos, it is easy to verify that only valid keys are there, like `assert(array_sum(array_map(function ($key) use ($options) { return array_key_exists($key, $options); }, $allowed_keys)) == count($options));`

I'd like to note that we (amp) added the keep_alive relatively late and one of the points were signal handler timers (pcntl_signal_dispatch each second for example) … this is only needed if you want to properly shutdown something. [Also allows for memory leak checking at the C level btw.] … I have a few cronjobs which use Artax to fetch in parallel … and then it should simply die at the end. I really don't want to die out of processes because unfinished cronjobs.
Possibly that need never occurred for Icicle or React; I don't know.

Regarding onEmpty - this will only fire if we don't have any timers etc. left. Doesn't make much sense here… The point after all is to have support for a case where we actually have timers left. (but have no real point where to stop them)

**Comment by bwoebi** [comment]
Sure, you can work around these issues if you maintain a counter whether all jobs are done and then call `stop()` - but I don't consider that a very elegant approach though.

**Comment by AndrewCarterUK** [comment]
Point taken regarding `onEmpty()`.

ping @trowski and @WyriHaximus for the Icicle and ReactPHP opinion on watcher options

**Comment by sagebind** [comment]
> Array options are the cleanest way though if we ever want to have more options.

I agree with @bwoebi here; there's no way you can account for the many vendor-specific options that may need to take into effect.

**Comment by trowski** [comment]
Vendor specific options arrays defeat the purpose of interoperability. If code depends on an option provided by a certain loop implementation, then we're back to the same place we are now where libraries only work with a specific event loop.

Icicle's event loop automatically stops when no _referenced_ watchers are pending in the loop. Watchers can be be referenced and unreferenced. All but signal watchers are referenced when created. In addition to `enable()`, `disable()`, and `cancel()`, `reference()` and `unreference()` methods could be added for referencing/unreferencing watchers. This is the strategy Node.js uses.

**Comment by kelunik** [comment]
Vendor specific options shouldn't be allowed there. Nowhere.

But an options array is way better than any boolean parameter. Booleans shouldn't be used as parameters except for pure setters.

**Comment by AndrewCarterUK** [comment]
Are we okay with removing array options and using `reference()` and `unreference()` in addition to `enable()`, `disable()` and `cancel()`?

It seems to me that this would be the cleanest solution and allow for more flexibility than locking this option in initially.

**Comment by kelunik** [comment]
I'm not that happy with the names `reference` / `unreference`, but don't have a suggestion right now.

If we go that route, watchers should be referenced by default.

**Comment by AndrewCarterUK** [comment]
What about visible and invisible?

**Comment by bwoebi** [comment]
I have no idea what visibility would mean in context of a reactor?

Referencing is definitely better …

On that subject, I wonder whether we really want to unreference? IMHO, it should be clear at setting time whether the watcher is keep_alive or not. Which is done best with an options array directly when setting.

**Comment by AndrewCarterUK** [comment]
I'm just throwing out names really, I don't have any problem with 'reference' and 'unreference'.

I'm strongly against the options array for the reasons stated. I would also suggest that it would be unlikely to go down well with the voting members in the FIG from a quality control perspective for the same reasons.

Providing room for future flexibility isn't an argument in favour. We can't add to this once it is standardised as it would be a break in compatibility. It would be better being a default '$keepAlive' parameter if this is the only option that we can think of and we don't want the extra methods.

**Comment by kelunik** [comment]
I think we can close this since it's already covered now with `reference` / `unreference`.

**Comment by kelunik** [comment]
Missed `cb_data`, if we don't want an options array, we might as well just use it as a new parameter.

**Comment by trowski** [comment]
I actually thought watchers had a `$cbData` argument, I'm surprised to see that they do not. This is an oversight on our part, since this argument is very helpful for implementing watcher callbacks without having to declare so many closures with `use` statements.

**Comment by WyriHaximus** [comment]
@AndrewCarterUK sorry missed your ping :/

Yes :+1:  on callback data, it would simplify passing data to a callback. Another option one could use is bind but that shouldn't be the default promoted way to go

**Comment by bwoebi** [comment]
Now really done with watcher data parameter in #35. Closing.

**Comment by rdlowrey** [comment]
Sorry for the late bump here. I've been preoccupied with work for the last year.

The problem with this design is it is not amenable to future API changes; it's brittle. The original reason why I put an options array in the v1 version of amphp is because it allows any one of many useful vendor-specific features. Callback data is just one of these useful features. I immediately ran into another missing functionality yesterday as soon as I tried porting some of the existing amp libs to use this interop spec. Namely creating a watcher that's disabled at creation time.

Consider the case where a client connects to my server and I _know_ I will need a watcher but just not now. I'm left doing one of two terrible things:
1. Check every time I need a writability watcher if one has yet been created. If so, enable it. If not, create one. Then repeat _every time_ I need to write to the socket.
2. Create the write watcher immediately upon client connect and then immediately call `disable()` on it to avoid the above problem (seems really spurious  and silly).

Now with the options API previously exposed in amphp I'm able to do the equivalent of this which  I'm prevented from doing with the interop version:

``` php
Loop::onWritable($socket, $callback, $options = [
    "callbackData" => $foo, // hey this still works  \o/
    "enableNow" => false, // don't enable the watcher right now to avoid  1 & 2 above
]);
```

Compare that to the option I'm left with if we stick with this more brittle API:

``` php
$watcherId = Loop::onWritable($socket, $callback, $foo);
// immediately pretend I didn't actually want to do that
// add extra fcall because of an inadequate creation API
Loop::disable($watcherId);
```

Callback data is just one of multiple useful facets of this  parameter. The ability to create a pre-disabled watcher is another. There are almost certainly more that we have yet to even think about. The only API that allows for future expansion without adding new parameters ad infinitum (one of the worst features of the PHP language's native functions BTW) is a trailing array of options.

Creating brittle APIs that don't account for future change about which we currently haven't even considered is how we end up with the kind of hateful APIs that litter the core of PHP ([canonical example](http://php.net/manual/en/function.stream-socket-client.php)). I really don't want to make that mistake here.

Yes, generally untyped array params are suboptimal, but IMO it's the correct option in this particular case.

I'm not willing to fall on the sword for this one, but to me it just seems like a more extensible and forward-looking API for this use-case. When designing APIs we should never assume we already know everything we'll ever need for the future (spoiler alert: we almost never do).

Anyway, I'll immediately disable watchers after creating them for now because the existing API doesn't prevent me from doing what I want. Just want to reopen debate on this to see if I can tempt  anyone else to take my viewpoint ;)

**Comment by kelunik** [comment]
> The original reason why I put an options array in the v1 version of amphp is because it allows any one of many useful vendor-specific features.

If you use vendor-specific features, it's not interoperable anymore. The watcher would still be enabled if you suddenly use React's loop.

> ridiculous option

I don't think it's ridiculous.

Given that, you might be right that we should design an API that's extensible. Still, we should also aim to prevent vendors to offer vendor-specific options.

**Comment by rdlowrey** [comment]
> I don't think it's ridiculous.

Agree to disagree. Though I can live with the extra call to disable it.

> Given that, you might be right that we should design an API that's extensible. Still, we should also aim to prevent vendors to offer vendor-specific options.

There's nothing un-interoperable about saying, _"we allow an options array to all for the possibility of future API expansion for which all vendors must support e.g. `callbackData` and `enableNow` keys"_ ... that's all I'm suggesting.

This is the  same reason why every single transport protocol ever created uses reserved bits ... because we don't know  what we might need in the future and we shouldn't wall off those possibilities.

**Comment by kelunik** [comment]
> There's nothing un-interoperable about saying, "we allow an options array to prevent future API expansion for which all vendors must support e.g. callbackData and enableNow keys" ... that's all I'm suggesting.

The point is that projects CAN support other options, but MOST NOT, otherwise it's un-interoperable.

**Comment by sagebind** [comment]
I don't think that an `options` is a _terrible_ idea, but if you are going to code against this standard alone, you cannot rely on _any_ vendor-specific features to be present or implemented, unless explicitly guaranteed by the standard.

If your code relied on  an Amp-specific `enableNow` feature, it would not work as expected when using with a different loop implementation, such as React. At worst case, your code completely breaks (which may or may not be the case for this specific case). You are essentially breaking the contract that your library is completely compatible with the standard.

**Comment by rdlowrey** [comment]
Yeah, I get what you guys are arguing. I'm suggesting we might want _standard_ options about which  we don't yet know in addition to the couple of valuable  ones we already do know about. At the same time we don't want more than three params as inputs into the function because: terrible.

**Comment by joshdifabio** [comment]
I think Daniel raises some great points, but I also think Andrew raised a valid point at the beginning of the discussion:

> Array options are something I'd like to avoid if at all possible.
> 
> It's can be difficult to detect typos when testing and I think it has the potential to encourage mistakes.

I've made typos in option arrays before and so have colleagues, and the silent failures have bitten us later on. Maybe, if option arrays were introduced, constants could be provided for the names of each of the various options? This would also provide a convenient place to document which options are part of the standard.

E.g.

``` php
class WatcherOption
{
    /** documentation here ... **/
    const ENABLE = 'enable';
    /** documentation here ... **/
    const CALLBACK_DATA = 'callbackData';

    const ON_WRITABLE_OPTIONS = [self::ENABLE, self::CALLBACK_DATA];
}

Loop::onWritable($socket, $callback, $options = [
    WatcherOption::CALLBACK_DATA => $foo,
    WatcherOption::ENABLE_WATCHER => false,
]);
```

Alternatively, there could be a separate class containing the options for each method, e.g. `OnWritableOptions`: this would mean more verbose code when we define option arrays, and more classes, but the benefit would be that we'd get nice method-specific autocompletion when specifying options:

``` php
class OnWritableOption
{
    /** documentation here ... **/
    const ENABLE = 'enable';
    /** documentation here ... **/
    const CALLBACK_DATA = 'callbackData';
}

Loop::onWritable($socket, $callback, $options = [
    OnWritableOption::CALLBACK_DATA => $foo,
    OnWritableOption::ENABLE_WATCHER => false,
]);
```

Alternatively the constants could all be shoved into `Loop`.

**Comment by sagebind** [comment]
If we want more options without adding a bunch of parameters, why not use some sort of object? We could avoid an array if we want to avoid arrays, and could have strong typing of options. You could even use some sort of builder, like this:

``` php
class WatcherOptions
{
    public static function new(): WatcherOptions; // constructor alias for chaining readability
    public function enabled(): WatcherOptions;
    public function disabled(): WatcherOptions;
    public function withCallbackData($data): WatcherOptions;
}

Loop::onWritable($socket, $callback, WatcherOptions::new()
    ->disabled()
    ->withCallbackData($foo)
    // maybe more options
);
```

**Comment by kelunik** [comment]
Dunno about the overhead if we use objects. What we could do is saying in the standard that unknown options MUST be rejected with an exception, so typos are always caught and vendors are less likely to introduce their own options.

**Comment by sagebind** [comment]
The object itself probably would not be stored though if you're concerned about memory. Implementors would get what they care about out of the object and then discard it.

**Comment by kelunik** [comment]
I'm more concerned about all the object creations and function calls for those low level functions that get called pretty often.

**Comment by rdlowrey** [comment]
I dunno lol ... I've kind of talked myself out of this already. We've been working on these APIs for nearly five years and the only two optional things of value I've encountered are:
- `callbackData`
- `enableNow`

I think my claims of impending API doom are probably a bit exaggerated. In the interest of keeping APIs as simple and broadly compatible as possible I think I can get behind calling `Loop::disable()` immediately after creation to emulate an `enableNow` option ...

Famous last words, but: _after all this time and with this many intelligent people contributing to the discussion I honestly don't think there are any other valid optional inputs to these functions_. And if  we come up  with some it would probably be enough to justify either a new major version or an additional API function to avoid breaking BC and supporting both.

Let's just keep it simple and not upset the apple cart as it currently exists.

**Comment by bwoebi** [comment]
Okay great. Closing this as resolved then.
